typescript/utils/src/addresses.ts

```diff
@@ -0,0 +1,240 @@
+    const isValid = address && ethersUtils.isAddress(address);
+    return !!isValid;
+  } catch (error) {
+    return false;
+  }
+}
+
+// Slower than isSealevelAddress above but actually validates content and checksum
+export function isValidSealevelAddress(address: Address) {
+  try {
+    const isValid = address && new PublicKey(address).toBase58();
+    return !!isValid;
+  } catch (error) {
+    return false;
+  }
+}
+
+export function isValidAddress(address: Address, protocol?: ProtocolType) {
+  return routeAddressUtil(
+    isValidEvmAddress,
+    isValidSealevelAddress,
+    false,
+    address,
+    protocol,
+  );
+}
+
+export function normalizeEvmAddress(address: Address) {
+  if (isZeroishAddress(address)) return address;
+  try {
+    return ethersUtils.getAddress(address);
+  } catch (error) {
+    return address;
+  }
+}
+
+export function normalizeSolAddress(address: Address) {
+  if (isZeroishAddress(address)) return address;
+  try {
+    return new PublicKey(address).toBase58();
+  } catch (error) {
+    return address;
+  }
+}
+
+export function normalizeAddress(address: Address, protocol?: ProtocolType) {
+  return routeAddressUtil(
+    normalizeEvmAddress,
+    normalizeSolAddress,
+    address,
+    address,
+    protocol,
+  );
+}
+
+export function areEvmAddressesEqual(a1: Address, a2: Address) {
+  return normalizeEvmAddress(a1) === normalizeEvmAddress(a2);
+}
+
+export function areSolAddressesEqual(a1: Address, a2: Address) {
+  return normalizeSolAddress(a1) === normalizeSolAddress(a2);
+}
+
+export function areAddressesEqual(a1: Address, a2: Address) {
+  const p1 = getAddressProtocolType(a1);
+  const p2 = getAddressProtocolType(a2);
+  if (p1 !== p2) return false;
+  return routeAddressUtil(
+    (_a1) => areEvmAddressesEqual(_a1, a2),
+    (_a1) => areSolAddressesEqual(_a1, a2),
+    false,
+    a1,
+    p1,
+  );
+}
+
+export function isValidEvmTransactionHash(input: string) {
+  return EVM_TX_HASH_REGEX.test(input);
+}
+
+export function isValidSolTransactionHash(input: string) {
+  return SEALEVEL_TX_HASH_REGEX.test(input);
+}
+
+export function isValidTransactionHash(input: string, protocol?: ProtocolType) {
+  return routeAddressUtil(
+    isValidEvmTransactionHash,
+    isValidSolTransactionHash,
+    false,
+    input,
+    protocol,
+  );
+}
+
+export function isZeroishAddress(address: Address) {
+  return ZEROISH_ADDRESS_REGEX.test(address);
+}
+
+export function shortenAddress(address: Address, capitalize?: boolean) {
+  if (!address) return '';
+  if (address.length < 8) return address;
+  const normalized = normalizeAddress(address);
+  const shortened =
+    normalized.substring(0, 5) +
+    '...' +
+    normalized.substring(normalized.length - 4);
+  return capitalize ? capitalizeAddress(shortened) : shortened;
+}
+
+export function capitalizeAddress(address: Address) {
+  if (address.startsWith('0x'))
+    return '0x' + address.substring(2).toUpperCase();
+  else return address.toUpperCase();
```

Comment:
interesting
alternatively
`toUpperCase().replace('X, 'x')`

Reply:
Could screw up for base58 addresses like on sealevel